jerry-core/ecma/base/ecma-helpers-conversion.c

```diff
@@ -917,227 +919,497 @@ ecma_number_to_int32 (ecma_number_t num) /**< ecma-number */
-                                              0ull,
-                                              0ull,
-                                              (fraction_uint64_m1) >> 32u,
-                                              ((fraction_uint64_m1) << 32u) >> 32u);
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_INIT (fraction_uint128_p1,
-                                              0ull,
-                                              0ull,
-                                              (fraction_uint64_p1) >> 32u,
-                                              ((fraction_uint64_p1) << 32u) >> 32u);
-
-  ecma_number_helper_binary_to_decimal (fraction_uint128, binary_exponent, &decimal_exp);
-  ecma_number_helper_binary_to_decimal (fraction_uint128_m1, binary_exponent_m1, &decimal_exp_m1);
-  ecma_number_helper_binary_to_decimal (fraction_uint128_p1, binary_exponent_p1, &decimal_exp_p1);
-
-  if (ECMA_NUMBER_CONVERSION_128BIT_INTEGER_IS_ZERO (fraction_uint128_m1))
+/**
+ * Normalize a floating point value
+ *
+ * @return normalized encoded_fp
+ */
+static encoded_float_t
+ecma_normalize_encoded_fp (encoded_float_t fp_value) /**< float value */
+{
+  JERRY_ASSERT (fp_value.fract != 0);
+
+  while (!(fp_value.fract & 0xFFC0000000000000ULL))
   {
-    decimal_exp_m1 = decimal_exp;
+    fp_value.fract <<= 10;
+    fp_value.exp -= 10;
   }
 
-  while (decimal_exp != decimal_exp_m1
-         || decimal_exp != decimal_exp_p1)
+  while (!(fp_value.fract & GRISU3_SIGN_BIT))
   {
-    while (decimal_exp > decimal_exp_m1
-           || decimal_exp > decimal_exp_p1)
-    {
-      ECMA_NUMBER_CONVERSION_128BIT_INTEGER_MUL_10 (fraction_uint128);
-      decimal_exp--;
-    }
-    while (decimal_exp_m1 > decimal_exp
-           || decimal_exp_m1 > decimal_exp_p1)
-    {
-      ECMA_NUMBER_CONVERSION_128BIT_INTEGER_MUL_10 (fraction_uint128_m1);
-      decimal_exp_m1--;
-    }
-    while (decimal_exp_p1 > decimal_exp
-           || decimal_exp_p1 > decimal_exp_m1)
-    {
-      ECMA_NUMBER_CONVERSION_128BIT_INTEGER_MUL_10 (fraction_uint128_p1);
-      decimal_exp_p1--;
-    }
+    fp_value.fract <<= 1;
+    --fp_value.exp;
   }
 
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_ADD (fraction_uint128_m1, fraction_uint128);
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_RIGHT_SHIFT (fraction_uint128_m1);
-
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_ADD (fraction_uint128_p1, fraction_uint128);
-  ECMA_NUMBER_CONVERSION_128BIT_INTEGER_RIGHT_SHIFT (fraction_uint128_p1);
+  return fp_value;
+} /* ecma_normalize_encoded_fp */
 
-  /* While fraction doesn't fit to integer, divide it by 10
-       and simultaneously increment decimal exponent */
-  uint64_t digits_min, digits_max;
+/**
+ * Convert ecma number to encoded_fp
+ *
+ * @return encoded_fp
+ */
+static inline encoded_float_t __attr_always_inline___
+ecma_number_to_encoded_fp (ecma_number_t value) /**< ecma number */
+{
+  encoded_float_t fp;
+  uint64_t *tmp = (uint64_t *) &value;
```

Comment:
This function is guarded with `CONFIG_ECMA_NUMBER_FLOAT64`, so this function cannot be called with float, but only with double.